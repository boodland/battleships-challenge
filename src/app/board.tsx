import BoardRow from './board-row';
import styles from './board.module.css';
import { Ship, columns, rows } from './utils';

export interface BoardProps {
  ships: Ship[];
  shots: boolean[];
  showShips: boolean;
  onShot: (position: number) => void;
}

const boardColumnsRow = (
  <div className={styles['columns']}>
    {columns.map((column) => (
      <span key={column} className={styles['text']}>
        {column}
      </span>
    ))}
  </div>
); 

function Board({ ships, shots, showShips, onShot }: BoardProps) {
  
  const boardCellRows = rows.map((position) => {
    return (
      <div key={position}>
        <span className={styles['text']}>{position + 1}</span>
        <BoardRow
          key={position}
          startPosition={position}
          ships={ships}
          shots={shots}
          showShips={showShips}
          onCellClick={onShot}
        />
      </div>
    );
  });

  return (
    <>
      {boardColumnsRow}
      {boardCellRows}
    </>
  );
}

export default Board;
