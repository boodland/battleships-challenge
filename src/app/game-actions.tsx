import { memo } from 'react';
import styles from './game-actions.module.css';

const CHECKBOX_LABEL = 'Show ships';

export interface GameActionsProps {
  showShips: boolean;
  onRestartGame: () => void;
  onToggleShowShips: () => void;
}

const GameActionsMemoized = memo(
  ({ showShips, onRestartGame, onToggleShowShips }: GameActionsProps) => {
    return (
      <div className={styles['actions']}>
        <button onClick={onRestartGame}>Restart</button>
        <div onClick={onToggleShowShips}>
          <input
            type="checkbox"
            checked={showShips}
            name={CHECKBOX_LABEL}
            readOnly
          />
          {CHECKBOX_LABEL}
        </div>
      </div>
    );
  }
);

export default GameActionsMemoized;
