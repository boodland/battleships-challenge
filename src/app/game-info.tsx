import styles from './game-info.module.css';
import { Ship, getShipsInfo, hasGameFinished } from './utils';

export interface GameInfoProps {
  ships: Ship[];
  shots: boolean[];
}

export function GameInfo({ ships, shots }: GameInfoProps) {
  const gameFinished = hasGameFinished(ships, shots);
  const shipsInfo = getShipsInfo(ships, shots).map((shipInfo, index) => {
    return (
      <div key={index} className={shipInfo.isShunk ? styles['shunk'] : ''}>
        {shipInfo.type}
      </div>
    );
  });

  return (
    <div className={styles['text']}>
      {shipsInfo}
      {gameFinished && 'Game Finished!'}
    </div>
  );
}

export default GameInfo;
