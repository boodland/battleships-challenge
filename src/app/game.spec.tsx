import { render } from '@testing-library/react';

import Game from './game';
import { BoardProps } from './board';
import { GameActionsProps } from './game-actions';
import { GameInfoProps } from './game-info';

const mocks = vi.hoisted(() => {
  return {
    hasGameFinished: vi.fn().mockReturnValue(false),
  }
})

vi.mock('./utils', () => {
  return {
    generateShips: () => [
      { type: 'Type 1', positions: [0, 1] },
      { type: 'Type 2', positions: [7, 8, 9] },
    ],
    initShots: () => Array(10).fill(false),
    hasGameFinished: mocks.hasGameFinished,
  };
});

vi.mock('./board', () => ({
  default: (props: BoardProps) =>
    `MockedBoard ${JSON.stringify(props)} onShot: ${props.onShot.name}`,
}));

vi.mock('./game-actions', () => ({
  default: (props: GameActionsProps) =>
    `MockedGameActions ${JSON.stringify(props)} onRestartGame: ${
      props.onRestartGame.name
    } onToggleShowShips: ${props.onToggleShowShips.name}`,
}));

vi.mock('./game-info', () => ({
  default: (props: GameInfoProps) => `MockedGameInfo ${JSON.stringify(props)}`,
}));

describe('Game', () => {
  it('should render successfully', () => {
    const { baseElement } = render(<Game />);
    expect(baseElement).toBeTruthy();
  });

  it('should display Board component', () => {
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });

  it('should display GameActions component', () => {
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });

  it('should display GameInfo component', () => {
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });

  it('should block board if game has finished', () => {
    mocks.hasGameFinished.mockReturnValue(true);
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });
});
