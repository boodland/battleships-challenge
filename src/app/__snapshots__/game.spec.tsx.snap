// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game > should block board if game has finished 1`] = `
<div>
  <div
    class="_game_b500ea"
  >
    <div
      class="_blocked_b500ea"
    >
      MockedBoard {"shots":[false,false,false,false,false,false,false,false,false,false],"ships":[{"type":"Type 1","positions":[0,1]},{"type":"Type 2","positions":[7,8,9]}],"showShips":false} onShot: onShot
    </div>
    <div
      class="_right-panel_b500ea"
    >
      MockedGameActions {"showShips":false} onRestartGame: restartGame onToggleShowShips: toggleShowShips
      MockedGameInfo {"shots":[false,false,false,false,false,false,false,false,false,false],"ships":[{"type":"Type 1","positions":[0,1]},{"type":"Type 2","positions":[7,8,9]}]}
    </div>
  </div>
</div>
`;

exports[`Game > should display Board component 1`] = `
<div>
  <div
    class="_game_b500ea"
  >
    <div
      class=""
    >
      MockedBoard {"shots":[false,false,false,false,false,false,false,false,false,false],"ships":[{"type":"Type 1","positions":[0,1]},{"type":"Type 2","positions":[7,8,9]}],"showShips":false} onShot: onShot
    </div>
    <div
      class="_right-panel_b500ea"
    >
      MockedGameActions {"showShips":false} onRestartGame: restartGame onToggleShowShips: toggleShowShips
      MockedGameInfo {"shots":[false,false,false,false,false,false,false,false,false,false],"ships":[{"type":"Type 1","positions":[0,1]},{"type":"Type 2","positions":[7,8,9]}]}
    </div>
  </div>
</div>
`;

exports[`Game > should display GameActions component 1`] = `
<div>
  <div
    class="_game_b500ea"
  >
    <div
      class=""
    >
      MockedBoard {"shots":[false,false,false,false,false,false,false,false,false,false],"ships":[{"type":"Type 1","positions":[0,1]},{"type":"Type 2","positions":[7,8,9]}],"showShips":false} onShot: onShot
    </div>
    <div
      class="_right-panel_b500ea"
    >
      MockedGameActions {"showShips":false} onRestartGame: restartGame onToggleShowShips: toggleShowShips
      MockedGameInfo {"shots":[false,false,false,false,false,false,false,false,false,false],"ships":[{"type":"Type 1","positions":[0,1]},{"type":"Type 2","positions":[7,8,9]}]}
    </div>
  </div>
</div>
`;

exports[`Game > should display GameInfo component 1`] = `
<div>
  <div
    class="_game_b500ea"
  >
    <div
      class=""
    >
      MockedBoard {"shots":[false,false,false,false,false,false,false,false,false,false],"ships":[{"type":"Type 1","positions":[0,1]},{"type":"Type 2","positions":[7,8,9]}],"showShips":false} onShot: onShot
    </div>
    <div
      class="_right-panel_b500ea"
    >
      MockedGameActions {"showShips":false} onRestartGame: restartGame onToggleShowShips: toggleShowShips
      MockedGameInfo {"shots":[false,false,false,false,false,false,false,false,false,false],"ships":[{"type":"Type 1","positions":[0,1]},{"type":"Type 2","positions":[7,8,9]}]}
    </div>
  </div>
</div>
`;
